### PR TITLE
Weij/multiple tokens user not provided

### PR DESCRIPTION
**Pull request by weijjia**
if silent request doesn't specify user and there are multiple tokens existed for the given app, and resource, we should fail.

### CODE REVIEW

File: adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenRequestTest.java [hunk 1]
```diff
@@ -398,6 +399,85 @@ public void testLocalSilentFailedBrokerSilentReturnErrorCannotTryWithInteractive
...
+                assertTrue(authenticationException.getCode().equals(ADALError.AUTH_FAILED_USER_MISMATCH));
+                latch.countDown();
+            }
+        });
+        latch.await();
+    }
+
+    @Test
+    public void testMutipleMRRTExistForTheSameApp() throws PackageManager.NameNotFoundException, InterruptedException {
+        final ITokenCacheStore cacheStore = getTokenCache(getExpireDate(MINUS_MINUITE), true, false, null);
```
Comment (jaredsburrows) [inline adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenRequestTest.java]: 👍 final

File: adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java [hunk 2]
```diff
@@ -258,6 +262,34 @@ void removeTokenCacheItem(final TokenCacheItem tokenCacheItem, final String reso
         Telemetry.getInstance().stopEvent(mTelemetryRequestId, cacheEvent,
                 EventStrings.TOKEN_CACHE_DELETE);
     }
+
+    boolean isMultipleRTsMatchingGivenAppAndResource(final String clientId, final String resource) {
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java]: I think we should look at merging these 2 functions together, pass null as the resource when you don't need resource.

File: adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java [hunk 3]
```diff
@@ -348,6 +380,13 @@ private boolean isUserMisMatch(final String user, final TokenCacheItem tokenCach
         return !user.equalsIgnoreCase(tokenCacheItem.getUserInfo().getDisplayableId()) 
                 && !user.equalsIgnoreCase(tokenCacheItem.getUserInfo().getUserId());
     }
+
+    private void throwIfMultipleATExisted(final String clientId, final String resource, final String user) throws AuthenticationException {
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java]: >throwIfMultipleATExisted [](start = 17, length = 24)

you should make this package scope and call it from AcquireTokenSilentHandler

File: adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java [hunk 4]
```diff
@@ -258,6 +262,34 @@ void removeTokenCacheItem(final TokenCacheItem tokenCacheItem, final String reso
...
+            if (tokenCacheItem.getAuthority().equalsIgnoreCase(mAuthority) && tokenCacheItem.getClientId().equalsIgnoreCase(clientId)
+                    && resource.equalsIgnoreCase(tokenCacheItem.getResource()) && !tokenCacheItem.getIsMultiResourceRefreshToken()) {
+                regularRTsMatchingRequest.add(tokenCacheItem);
+            }
+        }
+
+        return regularRTsMatchingRequest.size() > 1;
+    }
+
+    boolean isMultipleMRRTsMatchingGivenApp(final String clientId) {
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java]: >isMultipleMRRTsMatchingGivenApp [](start = 12, length = 31)

following making throwif* function pacakage scope you should make these 2 private.

File: adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java (hunk 2 above)
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java]: nah, that will make the lookup logic kinda messy.

File: adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java (hunk 3 above)
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java]: why?

File: adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java (hunk 4 above)
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java]: um, i cannot. They're used directly by AcquireTokenHandler. Let's talk tomorrow

**Pull request merged**